src/reflect/scala/reflect/io/ZipArchive.scala

```diff
@@ -123,37 +123,54 @@ final class FileZipArchive(file: JFile) extends ZipArchive(file) {
   lazy val (root, allDirs) = {
     val root = new DirEntry("/")
     val dirs = mutable.HashMap[String, DirEntry]("/" -> root)
-    val zipFile = try {
+
+    def openZipFile(): ZipFile = try {
       new ZipFile(file)
     } catch {
       case ioe: IOException => throw new IOException("Error accessing " + file.getPath, ioe)
     }
 
+    val zipFile = openZipFile()
     val enum    = zipFile.entries()
 
-    while (enum.hasMoreElements) {
-      val zipEntry = enum.nextElement
-      val dir = getDir(dirs, zipEntry)
-      if (zipEntry.isDirectory) dir
-      else {
-        class FileEntry() extends Entry(zipEntry.getName) {
-          override def getArchive   = zipFile
-          override def lastModified = zipEntry.getTime()
-          override def input        = getArchive getInputStream zipEntry
-          override def sizeOption   = Some(zipEntry.getSize().toInt)
+    try {
+      while (enum.hasMoreElements) {
+        val zipEntry = enum.nextElement
+        val dir = getDir(dirs, zipEntry)
+        if (zipEntry.isDirectory) dir
+        else {
+          // allows zipEntry to be garbage collected
+          val name = zipEntry.getName()
+          val time = zipEntry.getTime()
+          val size = zipEntry.getSize().toInt
+          class FileEntry() extends Entry(name) {
+            override def lastModified: Long = time // could be stale
+            override def input: InputStream = {
+              val zipFile  = openZipFile()
+              val entry    = zipFile.getEntry(name)
+              val delegate = zipFile.getInputStream(entry)
+              new FilterInputStream(delegate) {
+                override def close(): Unit = {
+                  delegate.close()
+                  zipFile.close()
```

Comment:
If someone _happened_ to be using this code to read zip entries and stream the contents into a file, or some other processing, that would be counter-productive. For the known user of this code in the compiler, it would be slightly counter productive, as the caller would just copy that Array immediately. We could of course change the caller (`AbstractFileReader.<init>`'s call to `AbstractFile.toByteArray`), but in the interests of keeping this change small I'd avoid starting down that path.